### DISCUSSION TITLE
Discussion on default domain behavior

### DISCUSSION THREAD

**Discussion by wfjsw** [shlinkio/discussions#84] [Ideas]
Currently, when a default domain is set, the `DropDefaultDomainFromRequestMiddleware` automatically erases the `domain` field if a link is created with a domain set explicitly if the domain filled equals to the default domain.

As a matter of personal preference, I dislike the behavior for various of reasons:

- The concept is weird, as a link on default domain would be available on all domains implicitly.
- I've changed the default domain once back then, and all link is simultaneously moved there, where it created mess. 
- I'm unable to create a link explicitly on the default domain.

So I went ahead and change the behavior to _explicitly fills the default domain into the `domain` field if no domain is specified_.

I want to know if it is only me who is suffering from this or it is a widespread pain.

**Comment by acelaya** [comment]
This behavior is set intentionally, to make Shlink backwards compatible to when it did not support multiple domains. Its behavior is documented here https://shlink.io/documentation/advanced/multiple-domains/

Working around this is pretty easy. Just set anything as the default domain, but never use it. Instead, always use specific domains. Short URLs created there will work for that domain only.

**Reply by wfjsw** [reply]
If I set the default domain to some random thing, it will end up showing in the frontend. Though I think it is a good practice to enable the REST API on a single domain only.

**Reply by acelaya** [reply]
Yeah, that's a good point.

In fact, I have to admit if I had to implement Shlink from scratch with multi-domain support, I would not do it the way it's now.

I could add a config option to let users decide how Shlink should behave. However this is a complex one to do, that would require a lot of tests to make sure nothing breaks.

I had an idea. I'm going to create a [new category](https://github.com/shlinkio/shlink/labels/%F0%9F%A5%87%20gold) of issues that I will implement if I reach my sponsorship goal.

I will add there the typical features that would be nice to have but I usually discard due to the complexity or effort required.
